### PR TITLE
[Impeller] Implement ui.Picture.toImage()

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/104718

Add a class `SnapshotController` for make snapshot, subclass `SnapshotControllerSkia` and `SnapshotControllerImpeller` to implement the function.  In addition, the API of `SnapshotDelegate:: MakeRasterSnapshot ` was modified,  the parameter `draw_callback` was replaced by `display_list`

I guess Impeller's implementation doesn't need to generate a raster image, right?

Also I'm not quite sure which tests I should add.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/common/rasterizer.h [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include <memory>
 #include <optional>
 
+#include "display_list/display_list_image.h"
```
Comment (ColdPaleLight) [inline shell/common/rasterizer.h:11]: The header file references here and elsewhere are a bit incorrect, I'll fix them later

**Comment by ColdPaleLight** [comment]
Would like to get some comments, cc @dnfield @chinmaygarde @jonahwilliams

File: shell/common/snapshot_controller.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                                            SkISize picture_size) = 0;
+
+  virtual sk_sp<SkImage> ConvertToRasterImage(sk_sp<SkImage> image) = 0;
+
+ protected:
+  explicit SnapshotController(const Delegate& delegate);
+  const Delegate& GetDelegate() { return delegate_; }
+
+ private:
+  const Delegate& delegate_;
```
Comment (dnfield) [inline shell/common/snapshot_controller.h:44]: FML_DISALLOW_COPY_AND_ASSIGN(SnapshotController)

File: shell/common/snapshot_controller_impeller.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include "flutter/flow/surface.h"
+#include "flutter/fml/trace_event.h"
+#include "flutter/impeller/display_list/display_list_dispatcher.h"
+#include "flutter/impeller/display_list/display_list_image_impeller.h"
+#include "flutter/impeller/geometry/size.h"
+#include "flutter/shell/common/snapshot_controller.h"
+
+namespace flutter {
+
+sk_sp<DlImage> SnapshotControllerImpeller::MakeRasterSnapshot(
```
Comment (dnfield) [inline shell/common/snapshot_controller_impeller.cc:17]: This method seems like it should be guarded by the GPU sync switch...

**Review by dnfield — COMMENTED** [review]
Overall looks good, would like Chinmay or Jason to get a chance to have a look at it too.

File: shell/common/snapshot_controller.h (hunk 2 above)
Comment (ColdPaleLight) [reply shell/common/snapshot_controller.h:44]: Done.

File: shell/common/snapshot_controller_impeller.cc (hunk 3 above)
Comment (ColdPaleLight) [reply shell/common/snapshot_controller_impeller.cc:18]: Done.

**Review by chinmaygarde — APPROVED** [review]

**Review by jason-simmons — APPROVED** [review]

**Comment by jonahwilliams** [comment]
Can we autosubmit this, or are there additional reviewers that should take a look?

**Comment by ColdPaleLight** [comment]
> Can we autosubmit this, or are there additional reviewers that should take a look?

I think we can autosubmit it.

**Pull request merged**